woocommerce/woocommerce-gateway-stripe
# UPE - Pre-orders support

Fixes #1820

## Changes proposed in this Pull Request:

- Convert pre-orders compatibility class into a trait.
- Refactor previous code so that we don't call `WC_Pre_Orders_*` methods directly outside the trait.
- Remove UPE disablement when pre-orders is active.
- Add support for UPE pre-orders.
- Fix a previous bug with SEPA pre-orders where Stripe would log out an error "missing payment_method".

## Testing instructions

- Make sure you have [WooCommerce Pre-Orders](https://woocommerce.com/products/woocommerce-pre-orders/) installed and active.
- For SEPA payments, make sure the store currency is set to EUR and you have webhooks configured: `stripe listen --forward-to "http://localhost:8082/?wc-api=wc_stripe"` and paste the webhook secret into the plugin settings.
- Run `npm run build:webpack`.
- Ensure there are no Stripe error logs at the end of the testing.

### Pre-orders with charge upon release.
1. Create a pre-order product with charge upon release:
  a. As a merchant, navigate to Products > Add new.
  b. Set the product title to: "Pre-order product with charge upon release".
  c. Under product data, add a regular price of 10.
  d. Select pre-orders.
  e. Check "Enable pre-orders for this product".
  f. Set an availability time in the future.
  g. Select when to charge: "Upon release".
  h. Click "Publish".
2. Paying with "new checkout experience" disabled:
  a. As a merchant, make sure "new checkout experience" is disabled under **WooCommerce > Settings > Payments > Stripe**.
  b. As a shopper, add the pre-order product to the cart and proceed to checkout.
  c. Pay with a new credit card.
  d. Make sure the order details page has a description at the top saying that you will be automatically charged for your order.
  e. As a merchant, navigate to **WooCommerce > Pre-Orders**.
  f. Select the recently purchased pre-order, go to **Bulk Actions > Complete > Apply**.
  g. Make sure the pre-order icon is changed to blue "Completed", and that the related order is "Processing".
  h. Please repeat the steps a-g with a saved credit card, and then with SEPA.
3. Paying with "new checkout experience" enabled:
  a. As a merchant, make sure "new checkout experience" is enabled under **WooCommerce > Settings > Payments > Stripe**.
  b. Repeat steps b-h from step 2.

### Pre-orders with charge upfront.
1. Create a pre-order product with charge upfront:
  a. As a merchant, navigate to Products > Add new.
  b. Set the product title to: "Pre-order product with charge upfront".
  c. Under product data, add a regular price of 10.
  d. Select pre-orders.
  e. Check "Enable pre-orders for this product".
  f. Set an availability time in the future.
  g. Select when to charge: "Upfront".
  h. Click "Publish".
2. Paying with "new checkout experience" disabled:
  a. As a merchant, make sure "new checkout experience" is disabled under **WooCommerce > Settings > Payments > Stripe**.
  b. As a shopper, add the pre-order product to the cart and proceed to checkout.
  c. Pay with a new credit card.
  d. Make sure the order details page has a description at the top saying that your order has been received, and no mentions of automatic payments in a future date.
  e. As a merchant, navigate to **WooCommerce > Orders** and make sure recently created order has a status of "Pre-ordered", and the order amount has already been charged.
  f. As a merchant, navigate to **WooCommerce > Pre-Orders**.
  g. Select the recently purchased pre-order, go to **Bulk Actions > Complete > Apply**.
  h. Make sure the pre-order icon is changed to blue "Completed", and that the related order is "Processing".
  i. Please repeat the steps a-h with a saved credit card, and then with SEPA.
3. Paying with "new checkout experience" enabled:
  a. As a merchant, make sure "new checkout experience" is enabled under **WooCommerce > Settings > Payments > Stripe**.
  b. Repeat steps b-i from step 2 also with Bancontact and iDEAL.

## Known issues

- Non-UPE pre-orders have a blank payment method name: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2104 (This one is new, but the behavior is present since before this change).
- The blocks checkout API still doesn't support pre-orders: https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1521.
- Pre-orders with charge upon release are not supported by payment request buttons.
